### ISSUE TITLE
Reduce Image Size

### ISSUE THREAD

**Issue by alensiljak** [dcycle/docker-ledger#4]
I was wondering if it would be possible to minimize the image size. The current one seems to be based on ubuntu and the total download is >500MB. 

Compared to ledger executable, which is <6MB, this is a significant overhead. 

Would it be possible to create an (official) image based on alpine linux image?

I noticed some difference between running a Windows version and the current GitHub version of Docker and would really like to have the same version running on different machines and platforms. Docker seems like a potential option but 500MB images are an overkill for that scenario.

I've seen some images (Dockerfiles) based on alpine but was unable to build them due to technical issues in the environment. I will try to resolve this and build an image but would still prefer if there was an official image, like this one.

Cheers and thanks for any consideration!

**Comment by alensiljak** [comment]
Hm, interestingly, I see alpine base image is 3MB and ubuntu latest is 32MB. While 10x is a significant relative difference, 30MB is still not much.
There's about 470MB of fat, so to speak. Boost library is fairly big but only a small fraction is used, I believe.

**Comment by alensiljak** [comment]
OpenSuse Tumbleweed [package](https://software.opensuse.org/package/ledger) is <800KB.
I know these are apples and oranges but I'm trying to figure out if it is possible to shave off >400MB from the image file.

**Comment by alensiljak** [comment]
Some extra info. I've built the image based on the Dockerfile in this repo. It certainly took some time and space. I've monitored the size of the image during the build process and it appears that before ledger cloning the size increases from ~20MB to 867MB.

![image](https://user-images.githubusercontent.com/462445/56419048-2f06bd80-6299-11e9-82c2-103e339c5456.png)

I'd like to point out that, according to an [answer on SO](https://stackoverflow.com/questions/24394243/why-are-docker-container-images-so-large)

> It is important to do that for each RUN, before the layer gets committed, or else deletes don't actually remove data.

there might be something that can be done in the various steps to clean up afterwards.

**Comment by alensiljak** [comment]
After successful build, these are the results: 1.76GB image!

![image](https://user-images.githubusercontent.com/462445/56419160-9ae92600-6299-11e9-9774-c901b67c1fbf.png)

Image history:
![image](https://user-images.githubusercontent.com/462445/56419210-c2d88980-6299-11e9-8ca9-2817f9cde513.png)

Images & Containers:
![image](https://user-images.githubusercontent.com/462445/56419234-d7b51d00-6299-11e9-95ec-32f8cfe5ca23.png)

**Comment by alensiljak** [comment]
Some pointers I'm reading currently: https://medium.com/travis-on-docker/microcontainers-tiny-portable-docker-containers-1507e3bf8688

**Comment by alensiljak** [comment]
Got an initial version going, thanks to @rgdoliveira!

https://gitlab.com/alensiljak/ledger-docker-alpine/

**Comment by alberto56** [comment]
Hey @MisterY thanks for this, great work! Are you open to joining forces and maintaining this image together? As next steps what do you think of:

* if you want I can make you a co-maintainer of this repo.
* we could put your code in this repo, in a new branch;
* once automated tests pass and the new branch is merged into master I can integrate this into my Jenkins server which rebuilds the image every week with the latest version and pushes to the docker hub at https://hub.docker.com/r/dcycle/ledger/tags using a time-based tag.
* not sure if we should use tag 2 for this or if it's similar enough that we should continue using tag 1 (for the image).

Thanks again!

**Comment by alensiljak** [comment]
Hi, @alberto56! I'd be happy to help out. This was an interesting exercise.
It would be useful to distinguish between alpine- and ubuntu-based images. Not sure how that works with tags yet. It would be worth looking how other packages utilize tags. Perhaps using `alpine` and `ubuntu` instead of 1 and 2 would work. Plus appending the date/time as you do now.

I've currently pushed the image to https://hub.docker.com/r/cick0/ledger-docker-alpine. This is a manually-built version, the result of testing the process.

I see that Docker Hub has integration with GitHub so perhaps you don't even need to link Jenkins in order to publish new images, but I haven't looked into that much.

Anyways, that's just a few ideas I have so far. I'm sure more will come on their own when the time is ripe. :)

**Comment by alberto56** [comment]
Cool, thanks. 

* Regarding integration between Docker and Github, that will not work in our case, because the Jenkins server is used to rebuild the image weekly, even if the Github code never changes. This makes sure (1) we always have the latest version of everything on the docker hub; (2) jenkins will throw an error in case our Dockerfile stops working with newer version of software, even if we have not modified the Github Dockerfile.
* Please feel free to open a pull request when your code is ready, at which time I'll give you access to this repo.
* I like the idea of using ubuntu, 1, and 1.2019-04-15-20-57-43-UTC for the Ubuntu version and perhaps alpine, 2, and 2.2019-04-15-20-57-43-UTC; that way, users who understand the difference between ubuntu and alpine can use the tag they want, and users who don't think in those terms would simply use dcycle/ledger:2 for the latest version. Once the alpine version is all tested, we can deprecate version 1/ubuntu and move to 2/alpine. At that point the [examples in the README.md file](https://github.com/dcycle/docker-ledger) and [the automated tests](https://github.com/dcycle/docker-ledger/blob/master/test.sh) would switch the using tag 2 or alpine.

**Comment by alensiljak** [comment]
OK, sounds good. My code is ready, I'd say. All I did was assemble a small Dockerfile. 
The real reliability might depend on the state of binary package for ledger-cli in Alpine Linux repo, as well as the dependencies it pulls, like Boost. As per [docs](https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management#Packages_and_Repositories), Boost is in Edge, while Ledger is in testing stage. 
I'll send the Dockerfile in a pull request.

**Comment by rgdoliveira** [comment]
@MisterY just a small correction. Boost package is in the main repository of Alpine, see[1]. So main repository comes enabled by default in Alpine apk (/etc/apk/repositories) but for ledger that is in testing, yeah, you need to add the "testing" repository in apk.

[1] https://pkgs.alpinelinux.org/package/edge/main/x86_64/boost

**Comment by alensiljak** [comment]
Hi, @rgdoliveira, that may be correct. What happened today is that I had issues building with only Testing repo added. Then I searched for `libboost`, which did not find anything. The search for `boost` showed version 1.69.0 (the same version I installed on my machine when I was building ledger-cli from source so I recognised it), but in Edge ~~repo~~ branch, see [here](https://pkgs.alpinelinux.org/packages?name=boost).
This may not be directly related but it prompted me to also add the Edge repository to the repositories file and after that the build succeeded.
In any case, if not necessary, I'd also prefer that the additional unstable repository does not get enabled, to avoid possibilities of any issues popping up unexpectedly. We can try that with an automated build.

**Comment by alensiljak** [comment]
Ah, OK, apologies. The repository is Main but the branch is Edge! I have mixed up the terminology, obviously.

**Comment by alensiljak** [comment]
Basically, [this section](https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management#Repository_pinning) got me on track. I tried pinning these, as suggested, but that did not work. So I just added both lines - edge and testing to the /etc/apk/repositories.

**Comment by rgdoliveira** [comment]
> Ah, OK, apologies. The repository is Main but the branch is Edge! I have mixed up the terminology, obviously.

Yes, Edge will be part of the next release.

**Comment by alberto56** [comment]
I will close this as https://github.com/dcycle/docker-ledger/pull/5 is now merged into master. It also builds (and tests) [10x faster](https://circleci.com/gh/dcycle/docker-ledger) !

This is now being tracked by my Jenkins server as well, which rebuilds the image with the latest version of the base image and Ledger CLI weekly, and it is on the Docker hub at https://hub.docker.com/r/dcycle/ledger/tags (as we can see the size went down from 500 to 23mb)

Thanks again @MisterY , I have also invited you to co-maintain this is you wish.

**Comment by alensiljak** [comment]
Yes, that's an improvement. Big thanks to the guys in Alpine Linux for the binary image!
